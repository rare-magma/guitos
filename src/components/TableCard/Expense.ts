import { ItemForm } from "../ItemForm/ItemForm";

export class Expense {
  items: ItemForm[] = [];
  total = 0;

  constructor(initializer?: Expense) {
    if (!initializer) return;
    if (initializer.items) this.items = initializer.items;
    if (initializer.total) this.total = initializer.total;
  }
}
