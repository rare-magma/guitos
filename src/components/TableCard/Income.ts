import { ItemForm } from "../ItemForm/ItemForm";

export class Income {
  items: ItemForm[] = [];
  total = 0;

  constructor(initializer?: Income) {
    if (!initializer) return;
    if (initializer.items) this.items = initializer.items;
    if (initializer.total) this.total = initializer.total;
  }
}
