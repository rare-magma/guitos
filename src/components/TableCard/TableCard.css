.btn-Expenses-plus-button {
  border: 1px solid var(--comment);
  color: var(--purple);
}

.btn-Expenses-plus-button:focus-visible,
.btn-Expenses-plus-button:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  color: var(--lightbgcolor);
}

.btn-Revenue-plus-button {
  border: 1px solid var(--comment);
  color: var(--highlight);
}

.btn-Revenue-plus-button:focus-visible,
.btn-Revenue-plus-button:hover {
  background-color: var(--highlight);
  border: 1px solid var(--highlight);
  color: var(--lightbgcolor);
}

.Expenses-card,
.Revenue-card {
  border: 1px solid var(--comment);
}

.Expenses-card-header,
.Revenue-card-header {
  border-bottom: 1px solid var(--comment);
}
