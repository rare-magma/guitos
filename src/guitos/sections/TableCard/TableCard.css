.btn-expenses-plus-button {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  color: var(--purple);
}

.btn-expenses-plus-button:focus-visible,
.btn-expenses-plus-button:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  color: var(--lightbgcolor);
}

.btn-revenue-plus-button {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  color: var(--highlight);
}

.btn-revenue-plus-button:focus-visible,
.btn-revenue-plus-button:hover {
  background-color: var(--highlight);
  border: 1px solid var(--highlight);
  color: var(--lightbgcolor);
}

.expenses-card,
.revenue-card {
  border: 1px solid var(--lightbgcolor);
}

.expenses-card-header,
.revenue-card-header {
  border-bottom: 1px solid var(--comment);
}

.btn-check:checked + .btn,
.btn.active,
.btn.show {
  box-shadow: none !important;
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-check + .btn:hover {
  box-shadow: none !important;
  border: 1px solid var(--cyan);
}

.btn-check:focus-visible + .btn,
.btn-check {
  box-shadow: none !important;
  color: var(--cyan);
  border: 1px solid var(--cyan);
}

.btn-outline-info.toggle {
  background-color: var(--bgcolor);
  border: 1px solid var(--lightbgcolor);
  border-radius: 0.375rem;
  color: var(--cyan);
}

.btn-outline-info.toggle:focus-visible,
.btn-outline-info.toggle:hover {
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-outline-info.toggle:disabled {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--lightbgcolor);
  border-radius: 0.375rem;
  color: var(--comment);
}
