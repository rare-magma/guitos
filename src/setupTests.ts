// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as matchers from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { randomUUID } from "node:crypto";
import { createElement } from "react";
import { afterEach, expect, vi } from "vitest";
import { Budget } from "./components/Budget/Budget";
import { ItemForm } from "./components/ItemForm/ItemForm";
import * as AppBudgetContext from "./context/BudgetContext";
import * as AppConfigContext from "./context/ConfigContext";
import * as AppGeneralContext from "./context/GeneralContext";

window.crypto.randomUUID = randomUUID;
global.URL.createObjectURL = vi.fn();

vi.mock("crypto", () => ({
  randomUUID: () => "035c2de4-00a4-403c-8f0e-f81339be9a4e",
}));

// silence recharts ResponsiveContainer error
vi.mock("recharts", async (importOriginal) => {
  const originalModule = await importOriginal();
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    ...originalModule,
    ResponsiveContainer: () => createElement("div"),
  };
});

// extends Vitest's expect method with methods from react-testing-library
expect.extend(matchers);

export const configContextSpy = vi.spyOn(AppConfigContext, "useConfig");
export const budgetContextSpy = vi.spyOn(AppBudgetContext, "useBudget");
export const generalContextSpy = vi.spyOn(
  AppGeneralContext,
  "useGeneralContext",
);

beforeEach(() => {
  budgetContextSpy.mockReturnValue(testBudgetContext);
  configContextSpy.mockReturnValue(testConfigContext);
  generalContextSpy.mockReturnValue(testGeneralContext);
});

// runs a cleanup after each test case (e.g. clearing jsdom)
afterEach(() => {
  cleanup();
});

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: vi.fn().mockImplementation((query: unknown) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // deprecated
    removeListener: vi.fn(), // deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

export const testBudget: Budget = {
  id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [{ id: 1, name: "expense1", value: 10 }],
    total: 10,
  },
  incomes: {
    items: [{ id: 2, name: "income1", value: 100 }],
    total: 100,
  },
  stats: {
    available: 90,
    withGoal: 80,
    saved: 10,
    goal: 10,
    reserves: 200,
  },
};
export const testBudgetClone: Budget = {
  ...testBudget,
  name: "2023-03-clone",
};

export const testBudget2: Budget = {
  id: "135b2ce4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-04",
  expenses: {
    items: [{ id: 1, name: "name", value: 50 }],
    total: 50,
  },
  incomes: {
    items: [{ id: 2, name: "name2", value: 200 }],
    total: 200,
  },
  stats: {
    available: 150,
    withGoal: 130,
    saved: 20,
    goal: 35,
    reserves: 30,
  },
};

export const testBigBudget: Budget = {
  id: "225c2de5-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [
      { id: 1, name: "name", value: 11378.64 },
      { id: 4, name: "name2", value: 11378.64 },
    ],
    total: 22757.28,
  },
  incomes: {
    items: [
      { id: 2, name: "name", value: 100.03 },
      { id: 3, name: "name2", value: 342783.83 },
    ],
    total: 342883.86,
  },
  stats: {
    available: 320126.58,
    withGoal: 148684.65,
    saved: 171441.93,
    goal: 50,
    reserves: 200,
  },
};

export const testJSONErrorBudget = `{
  id: "03123AAA5c2de4-00a4-403c-8f0e-f81339be9a4e",
  na2me: "2023-03",
  expens3es: {
    items: [{ id: "infinity", name: -1, value: "r" }],
    total: 10,
  },
  stats: {
    available: 0,
    withGoal: 0,
    saved: 0,
    goal: 10,
    reserves: 0,
  },
}`;

export const testCsv = `type,name,value
expense,rent,1000.00
expense,food,200.00
income,salary,2000.00
income,sale,100
goal,goal,10
reserves,reserves,0
`;

export const testCsvError = `type,name,value
expe2nse,rent,1000.00
expense,food,200.00,123,4
incomae,salary,2000.00
income,sale,100
goal,123,goal
goal,,goal,,,
reservaes,reserves,0
`;

export const testBudgetCsv = {
  id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [
      new ItemForm({ id: 0, name: "rent", value: 1000 }),
      new ItemForm({ id: 1, name: "food", value: 200 }),
    ],
    total: 1200,
  },
  incomes: {
    items: [
      new ItemForm({ id: 2, name: "salary", value: 2000 }),
      new ItemForm({ id: 3, name: "sale", value: 100 }),
    ],
    total: 2100,
  },
  stats: {
    available: 900,
    withGoal: 690,
    saved: 210,
    goal: 10,
    reserves: 0,
  },
};

export const budgetNameList = [
  {
    id: "035c2de4-01a4-403c-8f0e-f81340be9a4e",
    name: "2023-03",
  },
  {
    id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
    name: "2023-04",
  },
  {
    id: "036c2de4-00a4-402c-8f0e-f81339be9a4e",
    name: "2023-05",
  },
];

export const itemForm1 = new ItemForm({
  id: 1,
  name: "name1",
  value: 10,
});

export const itemForm2 = new ItemForm({
  id: 2,
  name: "name2",
  value: 100,
});

export const testCalcHist = [
  {
    id: `${testBudget.id}-Expenses-1`,
    itemForm: itemForm1,
    changeValue: 123,
    operation: "add",
  },
  {
    id: `${testBudget.id}-Expenses-1`,
    itemForm: {
      id: 1,
      name: itemForm1.name,
      value: 133,
    },
    changeValue: 3,
    operation: "add",
  },
];

export const testIntlConfig = { locale: "en-US", currency: "USD" };
export const testSpanishIntlConfig = { locale: "es-ES", currency: "EUR" };

export const testBudgetList = [testBudget, testBudget2, testBigBudget];

export const testBudgetNameList = [
  {
    id: testBudget.id,
    item: "",
    name: testBudget.name,
  },
  {
    id: testBudget2.id,
    item: "",
    name: testBudget2.name,
  },
];

export const setIntlConfigMock = vi.fn();
export const handleCurrencyMock = vi.fn();
export const testConfigContext = {
  intlConfig: testIntlConfig || undefined,
  setIntlConfig: setIntlConfigMock,
  currency: "USD",
  handleCurrency: handleCurrencyMock,
};

export const testSpanishConfigContext = {
  intlConfig: testSpanishIntlConfig,
  setIntlConfig: setIntlConfigMock,
  currency: "EUR",
  handleCurrency: handleCurrencyMock,
};

export const setBudgetMock = vi.fn();
export const setBudgetListMock = vi.fn();
export const setBudgetNameListMock = vi.fn();
export const undoMock = vi.fn();
export const redoMock = vi.fn();
export const testEmptyBudgetContext = {
  budget: undefined,
  setBudget: setBudgetMock,

  budgetList: [],
  setBudgetList: setBudgetListMock,

  budgetNameList: [],
  setBudgetNameList: setBudgetNameListMock,

  revenuePercentage: 0,
  past: [],
  future: [],
  needReload: true,
  undo: undoMock,
  redo: redoMock,
  canUndo: false,
  canRedo: false,
};

export const testBudgetContext = {
  budget: testBudget,
  setBudget: setBudgetMock,

  budgetList: testBudgetList,
  setBudgetList: setBudgetListMock,

  budgetNameList: testBudgetNameList,
  setBudgetNameList: setBudgetNameListMock,

  revenuePercentage: 10,
  past: [],
  future: [],
  undo: undoMock,
  redo: redoMock,
  canUndo: false,
  canRedo: false,
};

export const setNeedReloadMock = vi.fn();
export const setLoadingFromDBMock = vi.fn();
export const setErrorMock = vi.fn();
export const setCsvErrorsMock = vi.fn();
export const setJsonErrorsMock = vi.fn();
export const setShowErrorMock = vi.fn();
export const setNotificationsMock = vi.fn();
export const handleErrorMock = vi.fn();
export const testGeneralContext = {
  needReload: false,
  setNeedReload: setNeedReloadMock,
  loadingFromDB: false,
  setLoadingFromDB: setLoadingFromDBMock,
  error: "",
  setError: setErrorMock,
  csvErrors: [],
  setCsvErrors: setCsvErrorsMock,
  jsonErrors: [],
  setJsonErrors: setJsonErrorsMock,
  showError: false,
  setShowError: setShowErrorMock,
  notifications: [],
  setNotifications: setNotificationsMock,
  handleError: handleErrorMock,
};

const jsonErrors: AppGeneralContext.JsonError[] = [
  {
    errors:
      "SyntaxError: Expected ',' or '}' after property value in JSON at position 209",
    file: "123.json",
  },
];

const csvErrors: AppGeneralContext.CsvError[] = [
  {
    errors: [
      {
        type: "FieldMismatch",
        code: "TooFewFields",
        message: "Line 0: Too few fields: expected 3 fields but parsed 2",
        row: 0,
      },
    ],
    file: "123.csv",
  },
];

const error = "Thrown error";

export const testErrorGeneralContext = {
  ...testGeneralContext,
  error: error,
  showError: true,
};

export const testJsonErrorGeneralContext = {
  ...testGeneralContext,
  jsonErrors: jsonErrors,
  showError: true,
};

export const testCsvErrorGeneralContext = {
  ...testGeneralContext,
  csvErrors: csvErrors,
  showError: true,
};
