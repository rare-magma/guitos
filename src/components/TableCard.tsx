import { useState } from "react";
import { ItemForm } from "./ItemForm";
import {
  Card,
  Button,
  Row,
  Col,
  OverlayTrigger,
  Tooltip,
} from "react-bootstrap";
import ItemFormGroup from "./ItemFormGroup";
import {
  calcPercentage,
  calcTotal,
  intlFormat,
  round,
  userLang,
} from "../utils";
import { Expense } from "./Expense";
import { Income } from "./Income";
import { BsPlusLg } from "react-icons/bs";

interface TableCardProps {
  items: Income | Expense;
  revenueTotal: number;
  header: string;
  onChange: (table: Income | Expense) => void;
}

function TableCard({
  items: initialItems,
  revenueTotal,
  header: label,
  onChange,
}: TableCardProps) {
  const [table, setTable] = useState(initialItems);
  const [total, setTotal] = useState(calcTotal(table.items));
  const revenuePercentage = calcPercentage(total, revenueTotal);

  const addTable = (tableBeingEdited: Income | Expense) => {
    let newTable;
    if (label === "Revenue") {
      newTable = new Income();
    } else if (label === "Expenses") {
      newTable = new Expense();
    } else {
      throw new Error("Messed up table type");
    }

    const newItemForm = new ItemForm();

    let maxId;
    if (table.items.length !== 0) {
      maxId = Math.max(
        ...table.items.map((i) => {
          return i.id;
        })
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(calcTotal(newTable.items));
    onChange(newTable);
  };

  const removeTable = (toBeDeleted: ItemForm) => {
    let newTable;
    if (toBeDeleted.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.filter(
      (item: { id: number }) => item.id !== toBeDeleted.id
    );
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(calcTotal(newTable.items));
    onChange(newTable);
  };

  const handleChange = (item: ItemForm) => {
    let newTable;
    if (item.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.map((i) => {
      if (i.id === item.id) {
        return {
          id: item.id,
          name: item.name,
          value: isNaN(item.value) ? 0 : Number(item.value),
        };
      }
      return i;
    });
    newTable.total = calcTotal(newTable.items);

    setTable(newTable);
    setTotal(newTable.total);
    onChange(newTable);
  };

  return (
    <Card className={label + "-card"}>
      <Card.Header className={label + "-card-header"}>
        <Row>
          <Col>{label}</Col>
          <Col className="text-end">
            <OverlayTrigger
              placement="top"
              overlay={
                label === "Expenses" ? (
                  <Tooltip id={`tooltip-value-${label}`}>
                    {revenuePercentage}% of revenue
                  </Tooltip>
                ) : (
                  <></>
                )
              }
            >
              <div>{intlFormat(round(total * 100, 2), userLang)}</div>
            </OverlayTrigger>
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        {table.items?.map((item: ItemForm) => (
          <ItemFormGroup
            key={item.id}
            itemForm={item}
            costPercentage={calcPercentage(item.value, revenueTotal)}
            onChange={handleChange}
            onRemove={() => {
              removeTable(item);
            }}
          />
        ))}
        <div className="mt-3" />
        <div className="d-grid gap-2">
          <Button
            variant={label + "-plus-button"}
            size="sm"
            onClick={() => {
              addTable(table);
            }}
            name=""
            value=""
          >
            <BsPlusLg />
          </Button>
        </div>
      </Card.Body>
    </Card>
  );
}

export default TableCard;
