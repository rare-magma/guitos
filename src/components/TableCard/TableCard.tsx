import { useRef, useState } from "react";
import { ItemForm } from "../ItemForm/ItemForm";
import {
  Card,
  Button,
  Row,
  Col,
  OverlayTrigger,
  Tooltip,
} from "react-bootstrap";
import { BsPlusLg } from "react-icons/bs";
import Big from "big.js";
import { roundBig, calcTotal, calcPercentage, intlFormat } from "../../utils";
import ItemFormGroup from "../ItemForm/ItemFormGroup";
import { Expense } from "./Expense";
import { Income } from "./Income";
import { useConfig } from "../../context/ConfigContext";
import { useBudget } from "../../context/BudgetContext";

interface TableCardProps {
  items: Income | Expense;
  header: string;
  onChange: (table: Income | Expense) => void;
}

function TableCard({
  items: initialItems,
  header: label,
  onChange,
}: TableCardProps) {
  const [table, setTable] = useState(initialItems);
  const [total, setTotal] = useState(roundBig(calcTotal(table?.items), 2));
  const { budget, revenuePercentage } = useBudget();
  const inputRef = useRef<HTMLInputElement>(null);
  const { intlConfig } = useConfig();

  function addTable(tableBeingEdited: Income | Expense) {
    let newTable;

    if (label === "Revenue") {
      newTable = new Income();
    } else if (label === "Expenses") {
      newTable = new Expense();
    } else {
      throw new Error("Messed up table type");
    }

    const newItemForm = new ItemForm();

    let maxId;
    if (table && table.items.length !== 0) {
      maxId = Math.max(
        ...table.items.map((i) => {
          return i.id;
        }),
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = roundBig(calcTotal(newTable.items), 2);

    setTable(newTable);
    setTotal(roundBig(calcTotal(newTable.items), 2));
    onChange(newTable);
  }

  function removeTable(toBeDeleted: ItemForm) {
    let newTable;
    if (toBeDeleted.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.filter(
      (item: { id: number }) => item.id !== toBeDeleted.id,
    );
    newTable.total = roundBig(calcTotal(newTable.items), 2);

    setTable(newTable);
    setTotal(roundBig(calcTotal(newTable.items), 2));
    onChange(newTable);
  }

  function handleChange(item: ItemForm) {
    let newTable;
    if (item.constructor.name === "Income") {
      newTable = new Income();
    } else {
      newTable = new Expense();
    }
    newTable.items = table.items.map((i) => {
      if (i.id === item.id) {
        return {
          id: item.id,
          name: item.name,
          value: isNaN(item.value) ? 0 : Number(item.value),
        };
      }
      return i;
    });
    newTable.total = roundBig(calcTotal(newTable.items), 2);

    setTable(newTable);
    setTotal(newTable.total);
    onChange(newTable);
  }

  return (
    <Card
      key={`table-${label}-${intlConfig?.currency}`}
      className={label + "-card"}
    >
      <Card.Header className={label + "-card-header"}>
        <OverlayTrigger
          placement="top"
          overlay={
            label === "Expenses" ? (
              <Tooltip
                id={`tooltip-value-${label}`}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue
              </Tooltip>
            ) : (
              <></>
            )
          }
        >
          <Row>
            <Col>{label}</Col>
            <Col className="text-end fixed-width-font">
              <div>
                {intlConfig?.currency &&
                  intlFormat(roundBig(Big(total), 2), intlConfig?.currency)}
              </div>
            </Col>
          </Row>
        </OverlayTrigger>
      </Card.Header>
      <Card.Body>
        {table?.items?.map((item: ItemForm) => (
          <ItemFormGroup
            key={`${label}-${item.id}`}
            itemForm={item}
            label={label}
            costPercentage={
              budget ? calcPercentage(item.value, budget.incomes.total) : 0
            }
            inputRef={inputRef}
            onChange={handleChange}
            onRemove={() => {
              removeTable(item);
            }}
          />
        ))}
        <div className="mt-3" />
        <div className="d-grid gap-2">
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={`tooltip-new-itemformgroup`}
                style={{ position: "fixed" }}
              >
                add new item
              </Tooltip>
            }
          >
            <Button
              variant={label + "-plus-button"}
              aria-label={`add item to ${label}`}
              size="sm"
              onClick={() => {
                addTable(table);
                setTimeout(() => {
                  inputRef.current?.focus();
                }, 0);
              }}
              name=""
              value=""
            >
              <BsPlusLg />
            </Button>
          </OverlayTrigger>
        </div>
      </Card.Body>
    </Card>
  );
}

export default TableCard;
