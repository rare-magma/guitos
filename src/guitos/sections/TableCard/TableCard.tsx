import Big from "big.js";
import { produce } from "immer";
import { Reorder } from "motion/react";
import { useRef, useState } from "react";
import {
  Button,
  Card,
  Col,
  OverlayTrigger,
  Row,
  ToggleButton,
  Tooltip,
} from "react-bootstrap";
import { BsArrowsVertical, BsPlusLg } from "react-icons/bs";
import { intlFormat, roundBig } from "../../../utils";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { ItemFormGroup } from "../ItemForm/ItemFormGroup";
import "./TableCard.css";
import { BudgetCalculator } from "../../application/budgetCalculator";
import type { Budget } from "../../domain/budget";
import { BudgetItem } from "../../domain/budgetItem";
import type { Expenses } from "../../domain/expenses";
import type { Incomes } from "../../domain/incomes";

interface TableCardProps {
  header: "Revenue" | "Expenses";
}

// biome-ignore lint/style/noDefaultExport: lazy loading
export default function TableCard({ header: label }: TableCardProps) {
  const { budget, setBudget, revenuePercentage } = useBudget();
  const { userOptions } = useConfig();
  const inputRef = useRef<HTMLInputElement>(null);
  const [isDraggable, setIsDraggable] = useState(false);

  const isRevenue = label === "Revenue";
  const isExpense = label === "Expenses";
  const table = isExpense ? budget?.expenses : budget?.incomes;
  const total = table?.total ?? 0;

  function reorderTable(newOrder: BudgetItem[]) {
    if (!budget) return;
    const newState = produce((draft) => {
      if (isExpense) {
        draft.expenses.items = newOrder;
      } else {
        draft.incomes.items = newOrder;
      }
    }, budget);
    setBudget(newState(), true);
  }

  function handleTableChange(item: Incomes | Expenses) {
    if (!budget) return;
    const newState = produce((draft) => {
      if (isExpense) {
        draft.expenses = item;
      } else {
        draft.incomes = item;
      }
      draft.stats.available = roundBig(
        BudgetCalculator.available(draft as Budget),
        2,
      );
      draft.stats.withGoal = BudgetCalculator.availableWithGoal(
        draft as Budget,
      );
      draft.stats.saved = BudgetCalculator.saved(draft as Budget);
    }, budget);
    setBudget(newState(), true);
  }

  function addItemToTable(tableBeingEdited: Incomes | Expenses | undefined) {
    if (!tableBeingEdited) return;
    const tableHasItems = table && table.items.length !== 0;
    const newItemForm = BudgetItem.create();
    const newTable = isRevenue ? ({} as Incomes) : ({} as Expenses);
    let maxId: number;

    if (tableHasItems) {
      maxId = Math.max(
        ...table.items.map((i: BudgetItem) => {
          return i.id;
        }),
      );
    } else {
      maxId = 0;
    }

    newItemForm.id = Number.isNaN(maxId) ? 0 : maxId + 1;
    newItemForm.name = "";
    newItemForm.value = 0;

    newTable.items = tableBeingEdited.items.concat(newItemForm);
    newTable.total = roundBig(BudgetCalculator.itemsTotal(newTable.items), 2);

    handleTableChange(newTable);
  }

  return (
    <Card
      key={`table-${label}-${userOptions.currencyCode}-${table?.items.length}`}
      className={`${label}-card`}
    >
      <Card.Header className={`${label}-card-header`}>
        <OverlayTrigger
          placement="top"
          overlay={
            isExpense ? (
              <Tooltip
                id={`tooltip-value-${label}`}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue
              </Tooltip>
            ) : (
              // biome-ignore lint/complexity/noUselessFragments: react types need it
              <></>
            )
          }
        >
          <Row>
            <Col>{label}</Col>
            <Col className="text-end fixed-width-font">
              <div>{intlFormat(roundBig(Big(total), 2), userOptions)}</div>
            </Col>
          </Row>
        </OverlayTrigger>
      </Card.Header>
      <Card.Body>
        {table?.items && (
          <Reorder.Group
            axis="y"
            values={table.items}
            onReorder={reorderTable}
            as="div"
          >
            {table.items?.map((item: BudgetItem) => (
              <Reorder.Item
                key={item.id}
                value={item}
                as="div"
                dragListener={isDraggable}
              >
                <ItemFormGroup
                  key={`${label}-${item.id}-item-form-group`}
                  itemForm={item}
                  label={label}
                  costPercentage={
                    budget
                      ? BudgetItem.percentage(item.value, budget.incomes.total)
                      : 0
                  }
                  inputRef={inputRef}
                  userOptions={userOptions}
                />
              </Reorder.Item>
            ))}
          </Reorder.Group>
        )}
        <div className="mt-3" />
        <div className="d-flex gap-2">
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-new-itemformgroup"}
                style={{ position: "fixed" }}
              >
                add new item
              </Tooltip>
            }
          >
            <Button
              variant={`${label.toLowerCase()}-plus-button`}
              aria-label={`add item to ${label}`}
              size="sm"
              className="flex-grow-1"
              onClick={() => {
                addItemToTable(table);
                setTimeout(() => {
                  inputRef.current?.focus();
                }, 0);
              }}
              name=""
              value=""
            >
              <BsPlusLg aria-hidden={true} />
            </Button>
          </OverlayTrigger>
          <OverlayTrigger
            delay={250}
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-reorder-itemformgroup"}
                style={{ position: "fixed" }}
              >
                toggle reordering of items
              </Tooltip>
            }
          >
            <ToggleButton
              id={`toggle-reorder-${label}`}
              aria-label={`reorder items in ${label}`}
              type="checkbox"
              variant="outline-info toggle"
              value={1}
              size="sm"
              onClick={() => setIsDraggable(!isDraggable)}
              checked={isDraggable}
            >
              <BsArrowsVertical aria-hidden={true} />
            </ToggleButton>
          </OverlayTrigger>
        </div>
      </Card.Body>
    </Card>
  );
}
