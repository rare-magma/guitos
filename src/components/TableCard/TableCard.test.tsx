import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import { testBudget } from "../../setupTests";
import TableCard from "./TableCard";

describe("TableCard", () => {
  const onChange = vi.fn();
  const comp = (
    <TableCard
      items={testBudget.expenses}
      header={"Expenses"}
      onChange={onChange}
    />
  );

  beforeEach(() => {
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });
  it("renders initial Expenses state", () => {
    expect(screen.getByDisplayValue("expense1")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$10")).toBeInTheDocument();
  });

  it("renders initial Revenue state", () => {
    render(
      <TableCard
        items={testBudget.incomes}
        header={"Revenue"}
        onChange={onChange}
      />,
    );
    expect(screen.getByDisplayValue("income1")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$100")).toBeInTheDocument();
  });

  it("triggers onChange when user changes input", async () => {
    await userEvent.type(screen.getByDisplayValue("expense1"), "change name");

    expect(onChange).toBeCalledWith({
      items: [
        {
          id: 1,
          name: "expense1change name",
          value: 10,
        },
      ],
      total: 10,
    });
    expect(screen.getByDisplayValue("expense1change name")).toBeInTheDocument();

    await userEvent.type(screen.getByDisplayValue("$10"), "123");

    expect(onChange).toBeCalledWith({
      items: [
        {
          id: 1,
          name: "expense1change name",
          value: 123,
        },
      ],
      total: 123,
    });
    expect(screen.getByDisplayValue("$123")).toBeInTheDocument();
  });

  it("triggers onChange when user adds new item", async () => {
    await userEvent.click(
      screen.getByRole("button", { name: "add item to Expenses" }),
    );

    expect(onChange).toBeCalledWith({
      items: [
        {
          id: 1,
          name: "expense1change name",
          value: 123,
        },
        {
          id: 2,
          name: "",
          value: 0,
        },
      ],
      total: 123,
    });
  });

  it("triggers onChange when user deletes items", async () => {
    await userEvent.click(
      screen.getByRole("button", { name: "delete item 1" }),
    );
    await userEvent.click(
      screen.getByRole("button", { name: "confirm item 1 deletion" }),
    );

    expect(onChange).toBeCalledWith({
      items: [],
      total: 0,
    });
  });
});
