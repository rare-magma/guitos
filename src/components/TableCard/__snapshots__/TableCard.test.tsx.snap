// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableCard > matches snapshot 1`] = `
<TableCard
  header="Expenses"
  items={
    {
      "items": [
        {
          "id": 1,
          "name": "expense1",
          "value": 10,
        },
      ],
      "total": 10,
    }
  }
  onChange={[MockFunction spy]}
/>
`;
